Add TextColor property to BaseButton and make MathInputButton's color customizable

SymboLinker opened the pull request:
The main goal of this commit is to make keyboard colors customizable (per keyboard key) to fit an app's color theme(s). For example, a dark keyboard background with light button texts.

CSharpMath.Forms:
- Refactor BaseButton's constructor.
- Add TextColor property to BaseButton and a bindable TextColorProperty.
- Remove the predefined Red color from MathInputButton.InputToLaTeX(input) for "Backspace" and "Clear".

CSharpMath.Forms.Example (MathKeyboard.xaml):
- Use the TextColorProperty in the Style of "Backspace" and "Clear", setting it to Red.

Happypig375 commented:
Not a fan of use-once functions. @FoggyFinder?

Happypig375 commented:
Also would be nice if some new tests related to this change are added to CSharpMath.Xaml.Tests.

FoggyFinder commented:
> Not a fan of use-once functions. @FoggyFinder?

Yep, as well

SymboLinker commented:
OK, I will restore that and make some unit tests.

Op wo 16 sep. 2020 09:15 schreef FoggyFinder <notifications@github.com>:

> Not a fan of use-once functions. @FoggyFinder
> <https://github.com/FoggyFinder>?
>
> Yep, as well
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/verybadcat/CSharpMath/pull/164#issuecomment-693221353>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHVGU6X5NYQCY6PN6RKGSM3SGBQZ7ANCNFSM4RN5WWJA>
> .
>

Happypig375 commented on CSharpMath.Forms/Buttons.cs:
```diff
@@ -7,27 +7,56 @@ public abstract class BaseButton<TView, TPainter, TContent> : ImageButton
+        return stream;
+      }
+      return null;
+    });
```
This function can be inlined into the property setter as well by initializing `TextColor` in the constructor.

Happypig375 commented on CSharpMath.Forms/Buttons.cs:
```diff
@@ -7,27 +7,56 @@ public abstract class BaseButton<TView, TPainter, TContent> : ImageButton
+    public static readonly BindableProperty TextColorProperty =
+      BindableProperty.Create(nameof(TextColor), typeof(Color), typeof(BaseButton<TView, TPainter, TContent>), propertyChanged: (b, o, n) => {
+        var baseButton = (BaseButton<TView, TPainter, TContent>)b;
+        baseButton.TextColor = (Color)n;
```
Can be condensed into `=> ((BaseButton<TView, TPainter, TContent>)b).TextColor = (Color)n`.

SymboLinker commented:
Oops, I didn't read "CSharpMath.Xaml.Tests". 

While trying to find out where to write the unit tests, I was in doubt about the "Xaml" meaning. I thought that test project was meant for tests more related to the xml. I have created a test project "CSharpMath.Forms.Tests" in the "Xamarin.Forms" folder. Shall I move the tests and delete the newly created test project?

Happypig375 commented on CSharpMath.Forms.Tests/ButtonTests.cs:
```diff
@@ -0,0 +1,50 @@
+      foreach (var mathInputButton in TheMathInputButtons) {
+        // At the time of writing this test, vphatom = @"{\color{#00FFFFFF}{|}}" (Xamarin.Forms.Color.Transparent = "#00FFFFFF") is used as there is no \vphantom command yet.
+        // As soon as \vphantom has been implemented, the call .Replace(LatexHelper.vphantom, "") can be removed.
+        Assert.DoesNotContain(@"\color", mathInputButton.Content!.LaTeX!.Replace(LatexHelper.vphantom, ""));
```
I avoid using the null suppression operator. Should we allow it in tests or define a `NotNull` function which will raise `Xunit.Sdk.NotNullException` instead?

Happypig375 commented on CSharpMath.Forms/LatexHelper.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace CSharpMath.Forms {
+  public static class LatexHelper {
+    public static readonly string vphantom = SetColor("|", Color.Transparent);
```
`\hphantom` is for horizontal spacing while `\vphantom` is for vertical spacing. So `hphantom` would be more appropriate here.

SymboLinker replied:
The main purpose of the temporary "phantom" is, if I'm correct, to make sure that small symbols like the degree symbol are not streched. Using the vertical "|" symbol prevents that from happening. The comment in the code also says something about horizontal spacing, but I think that is not its main purpose. Besides, spacing between buttons can also be created in other ways that may be more appropriate - maybe the comment in the code is not correct.

Happypig375 replied:
Hmmm. What do you suggest on horizontal spacing?

SymboLinker replied:
I did not know about the \phantom command before working on this pull request, and I learned that the \phantom is a square (and \vphantom is a vertical line of zero width). If the \phantom is implemented and used instead of "|", the horizontal space will be increased. If a \vphantom is implemented, yes, then horizontal spacing must be managed another way. I think that managing the width of the button should be done by the developers using the buttons. Probably they will define the width of the buttons as a percentage of the screenwidth or something like that anyway. But if the intention is  "providing a helping hand in case the width is not managed by the developers", then probably the simple solution is the right one and using an invisible "|" is exactly that: it is related to the font size, which seems just perfect.
I have tried to find examples however where that minimum width is used. Is there an example in the example project that uses the minimum width that "|" causes?

SymboLinker replied:
Leaving the "|" width both responsibilities may be better than using \phantom or \vphantom. Developers that define width themselves don't notice it, and developers that do not define width are helped without asking. In that case the comment can be updated (remove the line that mentions \phantom).

Happypig375 replied:
See the Editor tab in CSharpMath.Forms.Example, I find that adding a tiny bit of horizontal space makes it look better

Happypig375 replied:
Well using `\phantom` is really the intention here. `\color` with transparent is a hack.

Happypig375 reviewed (commented):
Hmmm, if you create a new test project then make sure to include it in `CSharpMath.CrossPlatform.slnf` as well.

Happypig375 commented on CSharpMath.CrossPlatform.slnf:
```diff
@@ -27,7 +27,9 @@
       "CSharpMath.Forms/CSharpMath.Forms.csproj",
-      "CSharpMath.Forms.Example/CSharpMath.Forms.Example/CSharpMath.Forms.Example.csproj"
+      "CSharpMath.Forms.Example/CSharpMath.Forms.Example/CSharpMath.Forms.Example.csproj",
+	  "CSharpMath.Forms.Tests/CSharpMath.Forms.Tests.csproj"
+
```
```suggestion
	"CSharpMath.Forms.Tests/CSharpMath.Forms.Tests.csproj"
```

Happypig375 commented on CSharpMath.Forms.Tests/ButtonTests.cs:
```diff
@@ -23,28 +21,27 @@ public class ButtonTests {
+    [ClassData(typeof(TheMathInputButtons))]
+    public void AllMathInputButtonsHaveLatexContent(MathInputButton mathInputButton) {
+      Assert.False(string.IsNullOrEmpty(mathInputButton.Content?.LaTeX));
     }
```
```suggestion
    public void AllMathInputButtonsHaveLatexContent(MathInputButton mathInputButton) =>
      Assert.False(string.IsNullOrEmpty(mathInputButton.Content?.LaTeX));
```

Happypig375 commented on CSharpMath.Forms.Tests/ButtonTests.cs:
```diff
@@ -23,28 +21,27 @@ public class ButtonTests {
-      }
+    [Theory]
+    [ClassData(typeof(TheMathInputButtons))]
+    public void MathInputButtonsHaveTransparentBackgroundByDefault(MathInputButton mathInputButton) {
+      Assert.Equal(Color.Transparent, mathInputButton.BackgroundColor);
+    }
```
```suggestion
    public void MathInputButtonsHaveTransparentBackgroundByDefault(MathInputButton mathInputButton) =>
      Assert.Equal(Color.Transparent, mathInputButton.BackgroundColor);
```

Happypig375 commented on CSharpMath.Forms.Tests/TestHelpers/ComplexClassData.cs:
```diff
@@ -0,0 +1,11 @@
+using System.Collections.Generic;
+namespace CSharpMath.Forms.Tests.TestHelpers {
+  public abstract class ComplexClassData<T> : IEnumerable<object[]> where T : class {
```
This is just duplicating TheoryData of Xunit.

SymboLinker replied:
O, nice. I will delete this class and replace usings of it.

Happypig375 commented on CSharpMath.Forms.Tests/ButtonTests.cs:
```diff
@@ -23,28 +21,27 @@ public class ButtonTests {
+    public void MathInputButtonsHaveTransparentBackgroundByDefault(MathInputButton mathInputButton) {
+      Assert.Equal(Color.Transparent, mathInputButton.BackgroundColor);
+    }
+    public class TheMathInputButtons : TestHelpers.ComplexClassData<MathInputButton> {
```
I suggest `MemberData` instead of `ClassData` for less code.

Happypig375 replied:
Also, should use `MathKeyboardInput` as parameters since only xunit-serializable types will result in individual test cases instead of one single test case with multiple results.

Happypig375 commented on CSharpMath.Forms.Tests/TestHelpers/NotNull.cs:
```diff
@@ -0,0 +1,15 @@
+        throw new Xunit.Sdk.NotNullException();
+#nullable disable
+      return obj;
+#nullable restore
```
Why are the pragma directives here?

Happypig375 commented on CSharpMath.Forms.Tests/CSharpMath.Forms.Tests.csproj:
```diff
@@ -0,0 +1,6 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <ItemGroup>
+    <EmbeddedResource Include="..\CSharpMath.Rendering.Tests\ComicNeue_Bold.otf" Link="ComicNeue_Bold.otf" />
```
If you're not testing adding custom typefaces, then you should omit this line.

SymboLinker replied:
After deleting that line at least one of the unit tests failed. Maybe that was not the only thing I changed. I will re-check. (I thought it was related to creating the stream for setting the ImageSource of BaseButton and did not think about it really, I'm sorry.)

SymboLinker commented:
I really appreciate the feedback you are providing. This is the first pull request I've ever done, and I am surprised about the quick responses and thorough reviews, providing kind and helpful suggestions to make the code compact and clean. You also care about the code quality even more than I'd expected - commenting about a white line for example and using `=> bla;` instead of `=> { bla;}` - and I like that.

Happypig375 reviewed (commented):
It's because I am a victim of dead air in pull requests...
https://github.com/kostub/iosMath/pull/142
https://github.com/LayoutFarm/Typography/pull/192
https://github.com/fsharp/fslang-design/pull/442
https://github.com/MicrosoftDocs/visualfsharpdocs/pull/326
And you will most definitely face similar scenarios as well in the future.

Regarding code style, thanks - less code means fewer chances of bugs.

SymboLinker commented:
Maybe I should only have committed "Add MathInputButton_Command unit test and do the suggested cleaning 99417ce".
I thought "let's use the Keyboard getter to increase test coverage", but the intention of the test is more clear without it. I will revert the last two commits.

Happypig375 commented:
Try having tests that invoke https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.streamimagesource.stream?view=xamarin-forms#Xamarin_Forms_StreamImageSource_Stream of the `Source` property.

SymboLinker commented:
Hmm. The new tests fail in the github checks (not locally)...:

> The type initializer for 'SkiaSharp.SKImageInfo' threw an exception.

Happypig375 commented:
CSharpMath.Rendering.Tests test the SkiaSharp front end yet all images are drawn successfully. Weird.

Happypig375 commented:
Version mismatch? SkiaSharp is at 2.80.1 while SkiaSharp.Views.Forms is at 1.68.0.

SymboLinker commented:
When I update the package, another error shows up, which is also mentioned in https://github.com/mono/SkiaSharp/issues/1225.
A comment in that issue:

> It might be because the SkiaSharp.Views.Forms project does not actually have a netstandard implementation that can run in tests.

SymboLinker commented:
I've found a solution for using 2.80.2. I've downloaded the NuGetPackage, extracted it (zip file) and added the dll to the test project:
skiasharp.views.forms.2.80.2.nupkg\ref\netstandard2.0\SkiaSharp.Views.Forms.dll

I'll do some more checks and then I'll push.

SymboLinker commented:
Just clarifying: I've had two versions of the repository (with SkiaSharp.Views.Forms 1.68.0 and 2.80.2) that run the tests of CSharpMath.Forms.Tests successfully locally but fail in the GitHub checks.
I now continue my quest :)

SymboLinker commented:
OK, that didn't help.

Github Test  Error Message:
   System.AggregateException : One or more errors occurred. (The type initializer for 'SkiaSharp.SKImageInfo' threw an exception.)
---- System.TypeInitializationException : The type initializer for 'SkiaSharp.SKImageInfo' threw an exception.
-------- **System.DllNotFoundException : Unable to load shared library 'libSkiaSharp' or one of its dependencies.** In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: liblibSkiaSharp: cannot open shared object file: No such file or directory

SymboLinker commented:
Solving this NuGet/dll problem seems out of reach for me. I will revert my latest attempts. Options:
- We could remove or comment out the unit tests added in "Add MathButton unit tests that check that the image color is correcty set in the StreamImageSource (4223835)" and wait for the mono/SkiaSharp update to solve the bug (assuming that is the case).
-  I don't know on what device Github Tests run? I found a posting about configuring the machine on which the tests are run (I can't help any further with that...)
https://github.com/mono/SkiaSharp/issues/834#issuecomment-631479750
- Someone may be able to solve it another way...

Happypig375 commented:
> I don't know on what device Github Tests run?

The workflow runs on Linux (Ubuntu).

Happypig375 commented:
I'll follow up tomorrow, I think https://github.com/mono/SkiaSharp/releases/tag/v2.80.0#additional should be the solution.

Happypig375 commented:
I'm busy these few days, will come back to this a few more days later.

SymboLinker commented:
I will push another commit that will allow/stimulate customizing the keyboard keys even more.

from
![image](https://user-images.githubusercontent.com/32139898/94053642-e1ba4400-fdda-11ea-9fd2-fee66c33b7a7.png)

to (for example)
![image](https://user-images.githubusercontent.com/32139898/94053444-a4ee4d00-fdda-11ea-8085-b2153bb9f60c.png)

Happypig375 commented:
One of the squares is filled because it indicates where the cursor will be after pressing the button.

SymboLinker commented:
The main reason for not liking an empty square is that when I make the keyboard smaller vertically, the displayed borders of some empty sqaures are not of equal size. That may because of my computer screen, but still a reason to like to override the default:
![image](https://user-images.githubusercontent.com/32139898/94332129-a1bdb180-ffd2-11ea-96d9-dadb76e14ead.png)

Shall I remove the changes of the square in the Example project?

Or: make the empty square gray and leave the black square, via something like (using lumosity in case of TextColor changes):
```
  public class MyMathInputButton : MathInputButton {
    protected override string GetButtonDisplayLaTeX(MathKeyboardInput key) =>
      base.GetButtonDisplayLaTeX(key).Replace(@"\square", LatexHelper.SetColor("■", TextColor.WithLuminosity(0.7)));
    }
  }
```
![image](https://user-images.githubusercontent.com/32139898/94332476-fd3d6e80-ffd5-11ea-956e-5b782a772070.png)

Happypig375 commented:
If the button display is changed, the keyboard output should be changed too. @charlesroddie @FoggyFinder Is this change acceptable?

SymboLinker commented:
I would say that if the button display is customizable, the keyboard output need not change.
The keyboard output could become customizable too (that would be great actually), but I think those two need not be always the same.
- Imagine that someone would like to display `\times` on a button, but `\cdot` as output.
- Imagine that someone wants to make a calculator with `x^y` as the button-LaTeX...

FoggyFinder commented:
> If the button display is changed, the keyboard output should be changed too. @charlesroddie @FoggyFinder Is this change acceptable?

I think it would be better to move these changes into separate PR (or create an issue to discuss first).

But I personally don't like how new buttons look.

Happypig375 commented:
Personally I am okay with the fire button but don't like the idea of what you see not being what you get.

SymboLinker commented:
Just to clarify:
The point of discussion is whether MathInputButton.GetButtonDisplayLaTeX should be virtual or not.

In a discussion about that, the following comments...
- "But I personally don't like how new buttons look."
- "don't like the idea of what you see not being what you get."
...seem irrelevant because everyone can get what they want. The defaults stay the same.

A concern that is more important in my opinion:
Does this change have any negative technical implications for the development of this repository? For example: does it impose restrictions for classes beyond the MathInputButton class? (Does it violate the open–closed principle or does it make future changes harder because GetButtonDisplayLaTeX is not private anymore?)

I must say however, that this comment
"If the button display is changed, the keyboard output should be changed too."
is another example of feedback that I really appreciate. I'm just someone opening a pull request, and a proposal is taken [more serious than expected] / [beyond the intended scope].

SymboLinker commented:
For making the placeholder customizable in the output, I have created pull request https://github.com/verybadcat/CSharpMath/pull/167.

SymboLinker commented:
My proposal/plan for this PR after #167 has been merged/accepted:
- I'll sync/update this branch.
- I'll restore the placeholder appearances on the keyboard keys to what they were before.
- I'll add a button at the right of the Textbox "Enter keystrokes..." that says something like "Change appearance". Maybe only a single word is better, to keep more space for the textbox.
- The effect of clicking this new button that has the following effect **the first time**: the placeholder appearance changes on the keyboard buttons and in the output as well. (For example, using a LightGray full square instead of the currently used empty square with black border.)
- Clicking multiple times changes the appearance of keyboard keys again, for instance also changing the TextColor of all keys to brown, or changing the Clear button to a rubbish bin, and eventually (after a maximum of 5 clicks) the original appearance is visible again. This way, the defaults stay the same, but all added functionality is also available in the Example project.

SymboLinker commented:
I wanted to sync my fork in such a way that this pull request would not show the commit from #167, but alas. I chose "rebase", maybe "merge" would have done the trick. However, I also commited ddb5e4e and that is part of this pull request.

Happypig375 commented:
I'll try to solve Linux failures when I can allocate time...

SymboLinker commented:
I understand. Sometimes it's juggling priorities. I will stay tuned.

SymboLinker commented:
I would like to move on to other PRs but I also intend to avoid merge conflicts, so I'll wait with opening new PRs until this PR has been merged.

Proposal (I'll just do it and it can always be reverted):
- I remove the following two tests that have the Linux failures (DefaultButtonImageTextColorIsBlack, ButtonTextColorPropertyChangesImageColor).
- I will open a PR that re-adds the tests after this PR has been merged.

charlesroddie commented:
`[Fact(Skip="reason")]` ?

FoggyFinder reviewed (changes requested):
`TextColor` on buttons isn't reverted to `Black` after theme changed back to the first one.

SymboLinker commented:
> TextColor on buttons isn't reverted to Black after theme changed back to the first one.

This has never happened while I tested, but there is definitely something strange going on: if I deploy/run the UWP project several times and each time do the "Change appearance" test through all themes, for some deploys 1 or 2 buttons do not revert to Black when coming back to the first theme, and in other cases it just works. Testing now, I had the "+" button that stayed brown, in another deploy I had the "5" and "\leq" buttons that stayed brown...

Something to admit: I have not tested this for Android, because that has a compilation error:

> Can not resolve reference: `Validation`, referenced by `Xunit.SkippableFact`. Please add a NuGet package or assembly reference for `Validation`, or remove the reference to `Xunit.SkippableFact`.	CSharpMath.Forms.Example.Android

I don't know why "xunit"-related stuff (I thought they are for unit test only) can be found in Android's obj/project.assets.json and at the NuGet packages while they are not referenced in Android's csproj:
![afbeelding](https://user-images.githubusercontent.com/32139898/97620457-e5686880-1a21-11eb-9145-2eab1bfd7271.png)

SymboLinker commented:
My latest commit does not fix the issue FoggyFinder described.

Still investigating (at some moments I have time).

SymboLinker commented:
Ow, the unit tests think differently about removing the NullableColorBindablePropertyHelper. I will revert.

>     Xamarin.Forms.Xaml.XamlParseException : Position 1:97. Cannot assign property "PlaceholderActiveColor": Property does not exist, or is not assignable, or mismatching type between value and property

FoggyFinder commented on CSharpMath.Forms.Example/CSharpMath.Forms.Example/Controls/MathKeyboard.xaml.cs:
```diff
@@ -43,6 +42,26 @@ public enum Tab {
+
+    public void SetClearButtonImageSource(ImageSource imageSource) {
+      foreach(var button in ButtonGrids.SelectMany(grid => grid.Children)
+                            .Where(button => button is ImageSourceMathInputButton)
+                            .Cast<ImageSourceMathInputButton>()
```
```suggestion
                            .OfType<ImageSourceMathInputButton>()
```

SymboLinker commented:
@FoggyFinder I can't reproduce the bug anymore (of MathInputButtons not changing back to black when coming back to the first theme). I have tested this now, about 10 times, and 15 hours ago, about 20 times. Does it still occur for you?

If it does, the only thing I can think of to change, is listening to the following warning:

![image](https://user-images.githubusercontent.com/32139898/97779023-c1b93580-1b7b-11eb-95b1-ec5b504cf59b.png)

by defining `interface IButtonDraw { void ButtonDraw(); }` and using it on BaseButton's TextColor property:
```
    public Color TextColor { get => (Color)GetValue(TextColorProperty); set => SetValue(TextColorProperty, value); }
    public static readonly BindableProperty TextColorProperty = BindableProperty.Create(nameof(TextColor), typeof(Color), typeof(BaseButton<TView, TPainter, TContent>), Color.Black,
      propertyChanged: (b, o, n) => ((IButtonDraw)b).ButtonDraw());
```
(Casting b to BaseButton<TView, TPainter, TContent> would not work, because MathInputButton's override of the ButtonDraw method would not be executed.)

FoggyFinder commented:
> Does it still occur for you?

Yep, still the same

SymboLinker commented:
Core Tests results:
> Total tests: 937
     Passed: 933
    Skipped: 4

The test fail errors are not related to this PR.

Test Core fail reason:

> Error: Codecov failed with the following error: The process '/bin/bash' failed with exit code 2

Test iOS fail reason:

> 2020-11-01 08:20:56.474 CSharpMath.Ios.Tests[2654:25586] 	[FAIL] CSharpMath.Ios.Tests.Tests.MathDisplay(file: "ItalicAlignment", latex: "\\colorbox{yellow}P\\\\\\begin{array}{r}\\colorbox"...) : Assert.InRange() Failure\r\nRange:  (-1655.8 - 6525.8)\r\nActual: 6688
2020-11-01 08:20:56.474 CSharpMath.Ios.Tests[2654:25586] 		  at CSharpMath.Ios.Tests.Tests.Test (System.String directory, System.Action`1[T] init, System.String file, System.String latex) [0x0020a] in /Users/runner/work/CSharpMath/CSharpMath/CSharpMath.Ios.Tests/Tests.cs:54 

> 2020-11-01 08:20:55.897 CSharpMath.Ios.Tests[2654:25586] 	[FAIL] CSharpMath.Ios.Tests.Tests.MathInline(file: "ItalicAlignment", latex: "\\colorbox{yellow}P\\\\\\begin{array}{r}\\colorbox"...) : Assert.InRange() Failure\r\nRange:  (-1655.8 - 6525.8)\r\nActual: 6688
2020-11-01 08:20:55.897 CSharpMath.Ios.Tests[2654:25586] 		  at CSharpMath.Ios.Tests.Tests.Test (System.String directory, System.Action`1[T] init, System.String file, System.String latex) [0x0020a] in /Users/runner/work/CSharpMath/CSharpMath/CSharpMath.Ios.Tests/Tests.cs:54

SymboLinker commented:
@FoggyFinder I suspect that it is solved now via commit 486ac0f and I hope that calling ButtonDraw during each theme change in the example project is acceptable. Officially it is only needed to go from theme 1 to theme 2.

Happypig375 commented:
@FoggyFinder @SymboLinker I'll test this PR tomorrow

SymboLinker commented:
> @FoggyFinder @SymboLinker I'll test this PR tomorrow

@Happypig375 It's more than half a year later now. Could you please share why this PR has not been tested / has not gotten any response? I would like to know that, so that I can do better next time (in other public repositories).

Happypig375 commented:
It's just me focusing on other projects than this. I'll just merge this.

Happypig375 commented:
@FoggyFinder Just missing your approval

FoggyFinder reviewed (changes requested):
Color doesn't return back to Black in theme switching. I'll take a look soon and write my thoughts / questions

FoggyFinder commented:
> It's more than half a year later now. Could you please share why this PR has not been tested / has not gotten any response? I would like to know that, so that I can do better next time (in other public repositories).

I'm doing some testing. I totally newbie with Regex so getting help with those who know at least something (pretty much to my opinion). Expect feedback / commit with fix proposal or. at least. tests that are failing within 1-2 days.

SymboLinker commented:
@FoggyFinder 
- Why do you need RegEx? Is it because of the use of `Regex.Matches(latex, LaTeXSettings.PlaceholderActiveNucleus).Count > 1` in the MathInputButton? That is acutally a bug, I realize now! (The intention of that piece of code was to count the number of placeholders. I actually should not have used the Regex class there, because my intention was only to count the number of occurrences of a normal string. That's dangerous however, because LaTeX uses characters that may result in actual meaningful regular expressions. It was lazy to use that method instead of writing a simple "substring counting method" - I believe there are no such methods ready available.)
- Whenever I "think" I need to use regular expressions, I remind myself of [a joke about Regex](): _"Some people, when confronted with a problem, think "I know, I'll use regular expressions." Now they have two problems."_ (I try to avoid using Regex. Often, there are better/cleaner ways to solve problems without Regex.)
- If I do use it, I mostly Google a "C# Regex cheat sheet" (for example: [mikesdotnetting](https://www.mikesdotnetting.com/article/46/c-regular-expressions-cheat-sheet)). There are multiple versions of Regex. Some languages outside of .NET may have a slightly different syntax.
- Regex.Replace and Regex.Matches are examples of .NET methods that you can use to transform those Regex strings into doing something.

FoggyFinder commented:
Exactly. I thought "Aha. Now I realize why it behaves like this. And then - I can write simple Regex to fix it quickly. Well it ended with debug why a regex doesn't work properly in some case and then another regex and so on until the point when I totally no idea what the regex pattern means itself". 

Indeed I prefer not to use Regex and not only I'm very bad with them. But...you've been waiting for so long...my intention was to fix it ASAP,  merge in and then, probably, rewrite without Regex.

FoggyFinder commented:
The issue with display (like I saw brown color despite the theme is black) is the fact that `DrawButton` calls repeatedly in short period of time:

```csharp
        button.TextColor = color;
        button.PlaceholderRestingColor = placeholderRestingColor;
        button.PlaceholderActiveColor = placeholderActiveColor;
        button.ButtonDraw(/* If the LaTeXSettings change but the button's appearance properties don't, there's no event that causes the execution of this method. */);
```

So when we store "original" latex in a variable

        var originalLatexString = painter.LaTeX;

it might be previously mutated LaTeX that actually contains color etc attributes.

FoggyFinder commented:
I'm not a big fan of lock either but it was much easier than applying odd regex to prevent latex corruption.
Well, returning to black theme still shows light-red border in `frac`, ... but no visual glitch

SymboLinker commented:
Nice, using a lock in synchronous code is new for me. Thanks!

FoggyFinder commented on CSharpMath.Forms/MathInputButton.cs:
```diff
@@ -1,35 +1,72 @@
+            if (Input == MathKeyboardInput.Power || Input == MathKeyboardInput.Subscript) {
+              latex = latex.ReplaceFirstOccurrence(restingNucleus, coloredPlaceholderRestingNucleus);
+              latex = latex.ReplaceLastOccurrence(LaTeXSettings.PlaceholderActiveNucleus, coloredPlaceholderActiveNucleus);
+            } else if (LaTeXSettings.PlaceholderRestingNucleus == LaTeXSettings.PlaceholderActiveNucleus && latex.SubStringCount(LaTeXSettings.PlaceholderActiveNucleus) > 1
+                || restingNucleus != LaTeXSettings.PlaceholderActiveNucleus && latex.Contains(restingNucleus) && latex.Contains(LaTeXSettings.PlaceholderActiveNucleus)) {
```
One minor thing: can you add a comment to this compound condition? The meaning is not totally obvious. Have feeling (probably a wrong one) that some brackets are missing.

SymboLinker replied:
I totally agree that that piece of code is not very good readable. However, I'm not a fan of comments in code. Maybe I should refactor it? This could be commented code:
```
// unicode "\u25A1" is converted into latex "\square". Why only take this into account for the resting placeholder? I guess that the active placeholder will never be an empty square.
            var restingNucleus = LaTeXSettings.PlaceholderRestingNucleus.Replace("\u25A1", @"\square");
            var coloredPlaceholderRestingNucleus = LatexHelper.SetColor(restingNucleus, restingPlaceholderColor);
            var coloredPlaceholderActiveNucleus = LatexHelper.SetColor(LaTeXSettings.PlaceholderActiveNucleus, activePlaceholderColor);
            if (Input == MathKeyboardInput.Power || Input == MathKeyboardInput.Subscript) {
// In case of a power or subscript, the placeholder on the RIGHT is active on first appearance.
              latex = latex.ReplaceFirstOccurrence(restingNucleus, coloredPlaceholderRestingNucleus);
              latex = latex.ReplaceLastOccurrence(LaTeXSettings.PlaceholderActiveNucleus, coloredPlaceholderActiveNucleus);
            } else if (
// in ALL other cases of two placeholders,
// with the same nucleus (both a filled square, but with a different color)
LaTeXSettings.PlaceholderRestingNucleus == LaTeXSettings.PlaceholderActiveNucleus && latex.SubStringCount(LaTeXSettings.PlaceholderActiveNucleus) > 1
// or with a different nucleus, a filled square vs an empty square
                 || restingNucleus != LaTeXSettings.PlaceholderActiveNucleus && latex.Contains(restingNucleus) && latex.Contains(LaTeXSettings.PlaceholderActiveNucleus)) {
// the placeholder on the LEFT should be active on first appearance
              latex = latex.ReplaceFirstOccurrence(LaTeXSettings.PlaceholderActiveNucleus, coloredPlaceholderActiveNucleus);
              latex = latex.ReplaceLastOccurrence(restingNucleus, coloredPlaceholderRestingNucleus);
            } else {
// single placeholder should be active (let's not overwrite, but set the right color in one go)
              latex = latex.Replace(LaTeXSettings.PlaceholderActiveNucleus, coloredPlaceholderActiveNucleus);
            }
```
I realize that refactoring the code with same functional results, but wrong latex, is possible. I will add another unit test.

FoggyFinder replied:
Good enough for me. Yeah, I think refactoring makes sense but it can be done it later, not in this PR.

FoggyFinder reviewed (approved):
I've left one comment can you take a look on it? Anyway, approved.

SymboLinker commented:
Thanks @Happypig375 and @FoggyFinder!
